### PR TITLE
ConfDecoder: optionally decode with initial value

### PR DESCRIPTION
**Pull request by kitbellew**

### CODE REVIEW

**Comment by kitbellew** [comment]
@olafurpg @poslegm @gabro PTAL

File: metaconfig-core/jvm/src/test/scala/metaconfig/cli/TabCompletionSuite.scala
Comment (olafurpg) [inline metaconfig-core/jvm/src/test/scala/metaconfig/cli/TabCompletionSuite.scala:16]: Can we revert the diff in this file?

File: metaconfig-core/shared/src/main/scala/metaconfig/ConfDecoder.scala
Comment (olafurpg) [inline metaconfig-core/shared/src/main/scala/metaconfig/ConfDecoder.scala:44]: Can we move this to a separate file?

File: metaconfig-core/shared/src/main/scala/metaconfig/ConfDecoder.scala
Comment (olafurpg) [inline metaconfig-core/shared/src/main/scala/metaconfig/ConfDecoder.scala:45]: Exact name up for debate, but can we come up with something else than `apply`?
```suggestion
  def decoder(default: A): ConfDecoder[A]
```

**Comment by kitbellew** [comment]
@olafurpg @poslegm i will reply to code comments later but now I'm not so sure this approach actually works.

as I've come to realize, for nested structures, the trick that scalafmt uses is by listing decoders for each of its fields explicitly. would the macro implementing the factory be able to do the same?

even if, let's say, all the field factories were somehow implicitly available, how would we invoke them, too? not an expert on macros nor reflection.

@olafurpg, you asked if i could revert the diff to one of the test suites: i couldn't make that work, it was for some reason entering an infinite loop while instantiating the case class, and why that was affected by the change in the macro is not clear to me...

so, help? any suggestions?

File: metaconfig-core/jvm/src/test/scala/metaconfig/cli/TabCompletionSuite.scala
Comment (kitbellew) [reply metaconfig-core/jvm/src/test/scala/metaconfig/cli/TabCompletionSuite.scala:16]: done, figured it out

File: metaconfig-core/shared/src/main/scala/metaconfig/ConfDecoder.scala
Comment (kitbellew) [reply metaconfig-core/shared/src/main/scala/metaconfig/ConfDecoder.scala:44]: done

File: metaconfig-core/shared/src/main/scala/metaconfig/ConfDecoder.scala
Comment (kitbellew) [reply metaconfig-core/shared/src/main/scala/metaconfig/ConfDecoder.scala:45]: done

**Comment by kitbellew** [comment]
> @olafurpg @poslegm i will reply to code comments later but now I'm not so sure this approach actually works.

Well, maybe it does, see new patch.

> as I've come to realize, for nested structures, the trick that scalafmt uses is by listing decoders for each of its fields explicitly. would the macro implementing the factory be able to do the same?

The macro uses `Conf.getSettingOrElse` and actually passes both the type and the default, so it becomes easy to require an implicit field factory and use that.

So the only thing I needed to address is parsing derived types (`Option`, `Map`, `Seq`), for which the default provided is the instance of that type and not the generic. For that, `ConfDefault`, a wrapper around a type default value, is introduced.

> @olafurpg, you asked if i could revert the diff to one of the test suites: i couldn't make that work, it was for some reason entering an infinite loop while instantiating the case class, and why that was affected by the change in the macro is not clear to me...

Don't know why the infinite loop was happening, but I reversed the invocations and now it works.

File: metaconfig-core/shared/src/main/scala/metaconfig/ConfDecoder.scala
Comment (olafurpg) [inline metaconfig-core/shared/src/main/scala/metaconfig/ConfDecoder.scala:75]: For future PRs, please try to avoid unnecessary diffs like here since it makes reviewing the changes harder.

File: metaconfig-core/shared/src/main/scala/metaconfig/Conf.scala
Comment (olafurpg) [inline metaconfig-core/shared/src/main/scala/metaconfig/Conf.scala:27]: Breaking changes in the public Metaconfig API like in this diff here require a bump in the Scalafix major version. Custom Scalafix rules use this API to read configuration. For better or worse, I would like to avoid the situation where custom Scalafix rules that were compiled against Scalafix v0.9.14 stop working with Scalafix v0.9.15.

Is there a way to introduce `ConfDecoderFactory` without making it a superclass of `ConfDecoder`?

Comment (olafurpg) [reply metaconfig-core/shared/src/main/scala/metaconfig/Conf.scala:27]: I am BTW open to merge breaking changes in the `metaconfig.internal` package to unblock improvements in Scalafmt.

File: metaconfig-core/shared/src/main/scala/metaconfig/generic/package.scala
Comment (olafurpg) [inline metaconfig-core/shared/src/main/scala/metaconfig/generic/package.scala:16]: Commenting the review suggestion inline to elaborate on what I mean. Would it be possible to do something like this instead? The other decoder/encoder macros would then be implemented on top of this macro expansion. It would fix https://github.com/scalameta/metaconfig/issues/58 if we make the `default` argument by-name
```suggestion
  def deriveCodecFactory[T]: T => ConfCodec[T] =
```

Comment (kitbellew) [reply metaconfig-core/shared/src/main/scala/metaconfig/generic/package.scala:16]: this, or something very similar to this, is what didn't work in my first patch, leading to a circular dependency in one of the files which i worked around (without really understanding why it was failing) by moving some code into the companion object (and you requested to move back).

your suggestion will make one such usage simpler (but perhaps it could be emulated with an implicit conversion), but it would make chaining harder. what i am trying to allow, for instance, is the `scalafmt` case of presets (and also, if possible, avoiding creating an instance of a decoder for each instance of the class).

this is what `scalafmt` does today (all demonstrative, not exact):
```
class ScalafmtConfig(...) {
  implicit val decoderField1 = decoderForField1(field1)
  ...
  implicit val decoderFieldN = decoderForFieldN(fieldN)
  implicit val decoderThis = generic.deriveDecoder(this)
}
```
and then for presets I needed to modify it like this:
```
class ScalafmtConfig(...) {
  implicit val decoderField1 = decoderForField1(field1) // hopefully we don't forget presets here
  ...
  implicit val decoderFieldN = decoderForFieldN(fieldN)
  private val baseDecoderThis = generic.deriveDecoder(this)
  implicit val decoderThis = decoderForPresets(baseDecoderThis)
}
```
and this is what I would like to accomplish:
```
class Align(...)
object Align {
  implicit val decoder = ... // includes presets and possibly tokens.add handling
}
...
class ScalafmtConfig(...) {
 ...
}
object ScalafmtConfig {
  private val baseDecoder = generic.deriveDecoder(ScalafmtConfig.default)
  implicit val decoder = decoderForPresets(baseDecoderThis)
}
```
meaning, the decoder for a type will be defined with the type and every class using it will get it automatically. (and one instance, too.)

for that, special-case handling will need to be implemented as chaining of methods (or factories). presets handling looks like decoding `read(conf: Conf): Configured[(A, Conf)]` and then using that to apply regular generic decoder `read(conf: Conf, A: preset)`. and that needs to happen for each field, too, so must be inside `Conf.getSettingOrElse`.

hopefully, this makes some sense for you to comment further.

**Comment by kitbellew** [comment]
@olafurpg @poslegm please take a look at the current patchset, it should be fully binary compatible. i also tested with `scalafmt` and its presets logic, seems to work.

File: metaconfig-core/shared/src/main/scala/metaconfig/Conf.scala
Comment (olafurpg) [reply metaconfig-core/shared/src/main/scala/metaconfig/Conf.scala]: I'm OK with binary breaking changes in only 2.11 since Scalafix rules compile only against 2.12.

Adding a new abstract method to `ConfDecoder[A]` is binary breaking for clients that do `new ConfDecoder[A] { ... }`.

It's possible to enable MiMa to verify automatically what changes are binary compatible. So far this project hasn't had large API changes so I haven't worried about it

**Comment by olafurpg** [comment]
Can you elaborate on the exact motivation for this PR? Is the goal to avoid the pattern where we repeat inner `implicit val fieldN = ...` inside `ScalafmtConfig`? Or is the goal to enable some functionality that was not possible before?

In Scalafix we already achive factory-like functionality by declaring a method like this
```scala
def factory(default: ScalafixConfig): ConfDecoder[ScalafixConfig] =
  metaconfig.generic.deriveDecoder(default)
```

While this approach requires cumbersome boilerplate, it works OK. If the motivation of this PR is only to avoid boilerplate then I'm inclined to not merge it as I would like to avoid breaking changes in the API.

A bit of background: I'm not familiar with any Scala library that provides automatic decoders based on a default value. This was the primary motivation for creating Metaconfig in the first place (to support `style = ..` in Scalafmt). The current implementation may be a bit clunky, but it at least works.

**Comment by olafurpg** [comment]
I am not keen on adding a new `read()` method to the `ConfDecoder[T]` typeclass since it complicates the meaning of other operations like `map()`. A typeclass should only have one method.

**Comment by kitbellew** [comment]
@olafurpg thank you, i didn't know about type classes, they resemble template specializations in c++ which are unavailable in Java generics.

as to motivation, yes, it reduces a little bit of boilerplate but not much as the field types are not repeated (it might violate the DRY principle you occasionally refer to in your code comments).

primarily, it seems that i inadvertently derailed @poslegm #88 where he was adding useful functionality of merging values from conf with some input state (default or coming from earlier parsing of the same conf, like with presets).

in general, decoded scalafmt configuration is not fully defined from the initial value and simple generic parsing of conf, but a multistage process where output of each stage is a partial state and partial conf to be processed by the next stage.

for instance, first stage is applying presets and removing matched keys from conf. second is extracting incremental updates (.add and possibly .remove), then generic parsing and merging with prior steps. without moving the state through these decoders, hard to achieve cleanly.

as to binary compatibility. you're right, somehow everyone talks about adding methods with concrete implementation to a trait as breaking 2.11 but no one ever mentioned adding a method without an implementation (which is probably obvious, in hindsight).

**Comment by poslegm** [comment]
I still want to achieve default values support via `.add = ` in metaconfig. My first try on it with prioritized typeclasses has a smell, and I think that ideas about `ConfDecoderFactory` and `ConfDecoder.WithState` are very useful. I will dig into it when have time.

But as I understand this implementation still has the problem of past attempts: if user call `.map` on stateful decoder he got `ConfDecoder.WithoutState`. In ideal world all combinators should save previous data.

File: metaconfig-core/shared/src/main/scala/metaconfig/ConfDecoder.scala
Comment (poslegm) [inline metaconfig-core/shared/src/main/scala/metaconfig/ConfDecoder.scala:24]: I think in the direction of such API: what if user will call `map/flatMap` and other combinators on `ConfDecoderFactory` which will propagate state over transformations chain? And at the it he calls `.decoder` to generate final `ConfDecoder` with all context.

Comment (kitbellew) [reply metaconfig-core/shared/src/main/scala/metaconfig/ConfDecoder.scala:24]: the problem was that `state: => B` cannot be passed to underlying `ConfDecoder[A]`, and not to break compatibility, I added `mapExt` with covariance and `in/out` (maybe an implicit `CanBuildFrom` would also work).

read about type classes, very interesting (however, they don't require a single method and even allow inheritance):
* https://scalac.io/typeclasses-in-scala/ (also https://github.com/mpilquist/simulacrum)
* https://typelevel.org/cats/typeclasses.html
* https://alvinalexander.com/scala/fp-book/type-classes-101-introduction/

Comment (olafurpg) [reply metaconfig-core/shared/src/main/scala/metaconfig/ConfDecoder.scala:24]: My primary motivation to keep a single abstract method is that it makes it easier to reason about the meaning of methods like `map()`. It's the same reason I want to avoid multi-layer typeclasses because implicit resolution gets hairy when you have inheritance and you want to influence the priority of how typeclass instances get selected.

**Comment by olafurpg** [comment]
I am wondering if we can provide an out of the box solution for both the preset pattern + `.add` and extend the `deriveDecoder` macro to implement support for this without introducing changes to the public API 🤔 

> field types are not repeated (it might violate the DRY principle you occasionally refer to in your code comments).

When creating Metaconfig I was less concerned about repeated types more concerned about repeating field names in strings. A bit of boilerplate is OK IMO as long as you get compile-errors when you make a mistake. I still agree it would be nice if we can avoid the boilerplate of repeating inner fields to support presets.

**Comment by olafurpg** [comment]
To elaborate, we could provide a class like this in Metaconfig
```scala
case class Presets[T](values: Map[String, T])
```

That users can use like this
```scala
case class ScalafmtConfig(
  preset: Presets = Presets(Map("default" -> ScalafmtConfig.default, "align" -> ScalafmtConfig.align))
  align: Boolean = false)

object ScalafmtConfig {
  val default = ScalafmtConfig()
  val align = default.copy(align = true)
  val codec = metaconfig.generic.deriveDecoder(default)
}
```

The macro derived decoder would automatically replace the default value if the user defines a `preset`. I estimate this would require a fairly small change in the macro implementation around here https://github.com/scalameta/metaconfig/blob/aaab8feb0d7b22f6ae78dda3f291839537e2006b/metaconfig-core/shared/src/main/scala/metaconfig/internal/Macros.scala#L118

**Comment by olafurpg** [comment]
I am open to do something similar to solve the `.add/.remove` pattern, but I don't have a concrete proposal at the top of my head

**Comment by olafurpg** [comment]
BTW, I am open to introduce breaking changes. I am planning to do a Scalafix v1.0 release at some point in which I think it's OK to upgrade to a new stable version of Metaconfig.

**Comment by kitbellew** [comment]
> BTW, I am open to introduce breaking changes. I am planning to do a Scalafix v1.0 release at some point in which I think it's OK to upgrade to a new stable version of Metaconfig.

@olafurpg that makes it easier. i will, however, yield to @poslegm and his #92 as it was his idea to start with. i got carried away :)

**Comment by kitbellew** [comment]
@olafurpg here's a version which doesn't break binary compatibility and has only one method in the interface (aka, type class). ptal.

**Review by olafurpg — APPROVED** [review]
LGTM! Thank you for preserving binary compatibility 🙏 

it’s not ideal to have duplicated the ConfDecoder class but it’s the lesser evil IMO. If you prefer, you can put ConfDecoderEx class in an internal package so you can break the api in the future.

**Comment by kitbellew** [comment]
> LGTM! Thank you for preserving binary compatibility 🙏
> 
> it’s not ideal to have duplicated the ConfDecoder class but it’s the lesser evil IMO. If you prefer, you can put ConfDecoderEx class in an internal package so you can break the api in the future.

will do, thanks

**Pull request merged**
